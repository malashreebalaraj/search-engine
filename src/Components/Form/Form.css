.Form {
}
body {
  font: 16px;
  font-family: "Trebuchet MS", "Lucida Sans Unicode", "Lucida Grande",
    "Lucida Sans", Arial, sans-serif;
  height: 100vh;
  background-image: url(https://images.pexels.com/photos/1370298/pexels-photo-1370298.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=750&w=1260);
  background-size: cover;
  background-position: center;
}

.Form input {
  width: 350px;
  margin: 5px;
  position: relative;
  font-size: 16x;
  padding: 10px 5px;
  box-sizing: border-box;
}

.Form button {
  width: auto;
  margin: 5px;
}

.Form ul {
  width: 350px;
  list-style-type: none;
  text-align: left;
  /* margin: auto 30px auto 396px; */
  margin: auto auto auto auto;
  padding: 0;
  border: none;
  color: black;
  text-indent: 0.01px;
  box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.1), 0 2px 4px 1px rgba(0, 0, 0, 0.18);
  background-color: white;
  z-index: 999;
  position: relative;
}

.Form ul::before {
  content: "";
}

.Form li {
  padding: 10px 5px;
  cursor: pointer;
  position: relative;
  z-index: 999;
  margin: auto auto auto auto;
}

.Form li:hover {
  background-color: rgba(128, 128, 128, 0.2);
}
