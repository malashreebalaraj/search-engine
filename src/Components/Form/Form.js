import React, { Component } from "react";
import "./Form.css";
import indexData from "../../Search/IndexData";
import search from "../../Search/Search";
import Book from "../Book/Book";

const querySummary = indexData();

class Form extends Component {
  constructor(props) {
    super(props);
    this.bookResult = [];
    this.state = {
      searchInput: "",
      suggestions: [],
      result: [],
      resultIndex: 0,
    };
  }

  changeHandler = (e) => {
    var searchInput = e.target.value;
    let suggestions = [];
    let searchResult = [];
    if (searchInput.length > 0) {
      searchResult = search(searchInput, querySummary);
      for (var i = 0; i < searchResult.length; i++) {
        suggestions.push(searchResult[i][0]);
      }
    }
    this.setState({
      searchInput: e.target.value,
      suggestions: suggestions,
      result: searchResult,
    });
  };

  suggestionSelected(value, index) {
    this.setState(() => ({
      searchInput: value,
      suggestions: [],
      resultIndex: index,
    }));
  }

  renderSuggestions() {
    if (this.state.suggestions.length === 0) {
      return null;
    }
    return (
      <ul>
        {this.state.suggestions.map((item, index) => (
          <li key={index} onClick={() => this.suggestionSelected(item, index)}>
            {item}
          </li>
        ))}
      </ul>
    );
  }

  onSubmitHandler = (e) => {
    if (this.state.result.length === 0) {
      return null;
    }
    e.preventDefault();

    this.setState({
      bookResult: this.bookResult.push(
        this.state.result[this.state.resultIndex]
      ),
    });
  };

  render() {
    return (
      <form className="Form">
        <React.Fragment>
          <h1>
            <span
              style={{
                color: "black",
                background: "#FFFFFF7F",
              }}
            >
              Search Books
            </span>
          </h1>
          <input
            data-testid="search-box"
            type="text"
            placeholder="Enter your search here"
            value={this.state.searchInput}
            onChange={(e) => this.changeHandler(e)}
          />
          <button
            data-testid="button"
            value="submit"
            onClick={(e) => this.onSubmitHandler(e)}
          >
            Submit
          </button>

          {this.renderSuggestions()}

          {this.bookResult &&
            this.bookResult.map((book, index) => (
              <div>
                <Book key={index} result={book} />
              </div>
            ))}
        </React.Fragment>
      </form>
    );
  }
}

export default Form;
