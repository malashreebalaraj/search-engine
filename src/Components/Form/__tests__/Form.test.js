import React from "react";
import ReactDOM from "react-dom";
import Form from "./../Form";

import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import renderer from "react-test-renderer";

afterEach(cleanup);

test("renders Form component without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Form />, div);
});

test("renders search books header text", () => {
  const { getByText } = render(<Form />);
  const textElement = getByText(/search books/i);
  expect(textElement).toBeInTheDocument();
});

test("renders search input box", () => {
  const { getByTestId } = render(<Form />);
  const searchElement = getByTestId("search-box");
  expect(searchElement).toBeInTheDocument();
});

test("renders placeholder text in search input box", () => {
  const { getByPlaceholderText } = render(<Form />);
  const textElement = getByPlaceholderText(/enter your search here/i);
  expect(textElement).toBeInTheDocument();
  //   expect(textElement).toHaveTextContent(/enter your search here/i);
});

test("renders submit button", () => {
  const { getByTestId } = render(<Form />);
  const buttonElement = getByTestId("button");
  expect(buttonElement).toBeInTheDocument();
});

test("check button label text is Submit", () => {
  const { getByTestId } = render(<Form />);
  const buttonElement = getByTestId("button");
  expect(buttonElement).toHaveTextContent("Submit");
});

test("matches snapshot of Form component", () => {
  const tree = renderer.create(<Form />).toJSON();
  expect(tree).toMatchSnapshot();
});
