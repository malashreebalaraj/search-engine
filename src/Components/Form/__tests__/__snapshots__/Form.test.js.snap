// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot of Form component 1`] = `
<form
  className="Form"
>
  <h1>
    <span
      style={
        Object {
          "background": "#FFFFFF7F",
          "color": "black",
        }
      }
    >
      Search Books
    </span>
  </h1>
  <input
    data-testid="search-box"
    onChange={[Function]}
    placeholder="Enter your search here"
    type="text"
    value=""
  />
  <button
    data-testid="button"
    onClick={[Function]}
    value="submit"
  >
    Submit
  </button>
</form>
`;
